man/man1/lpq.1
.\" @(#)lpq.1 1.1 92/07/30 SMI; from UCB 4.3 BSD
.TH LPQ 1 "9 September 1987"
.SH NAME
lpq \- display the queue of printer jobs
.SH SYNOPSIS
.B lpq
[
.BI \-P printer
] [
.B \-l
] [
.B +
[
.I interval
] ] [
.I job#
\&.\|.\|.
] [
.I username
\&.\|.\|.
]
.SH DESCRIPTION
.IX  "lpq command"  ""  "\fLlpq\fP \(em display printer queue"
.IX  printer  "display queue"  ""  "\fLlpq\fP \(em display queue"
.IX  display "printer queue \(em \fLlpq\fP"
.IX  queue  "display printer"  ""  "\fLlpq\fP \(em display printer"
.LP
.B lpq
displays the contents of a printer queue.  It reports the status of
jobs specified by
.IR job# ,
or all jobs owned by the user specified by
.IR username .
.B lpq
reports on all jobs in the default printer queue when invoked with no
arguments.
.LP
For each print job in the queue,
.B lpq
reports the user's name, current position, the names of input
files comprising the job, the job number (by which it is
referred to when using
.BR lprm (1))
and the total size in bytes.
Normally, only as much information as will fit on one line
is displayed.  Jobs are normally queued on a first-in-first-out basis.
Filenames comprising a job may be unavailable, such as when
.B lpr
is used at the end of a pipeline; in such cases the filename field
indicates ``(standard input)''.
.LP
If
.B lpq
warns that there is no daemon present
(that is, due to some malfunction), the
.BR lpc (8)
command can be used to restart a printer daemon.
.SH OPTIONS
.TP 15
.BI \-P " printer"
Display information about the queue for the specified
.IR printer .
In the absence of the
.B \-P
option, the queue to the printer specified by the
.SM PRINTER
variable in the environment is used. If the
.SM PRINTER
variable isn't set, the queue for the default printer is used.
.TP
.BI \-l
Display queue information in long format; includes the name of
the host from which the job originated.
.HP
.BR \(pl [
.I interval
]
.br
Display the spool queue periodically until it empties.  This option
clears the terminal screen before reporting on the queue.  If an
.I  interval
is supplied,
.B lpq
sleeps that number of seconds in between reports.
.SH FILES
.PD 0
.TP 20
.B /etc/termcap
for manipulating the screen for repeated display
.TP
.B /etc/printcap
to determine printer characteristics
.TP
.B /var/spool/l*
spooling directory, as determined from printcap
.TP
.B /var/spool/l*/cf*
control files specifying jobs
.TP
.B /var/spool/l*/lock
lock file to obtain the currently active job
.PD
.SH "SEE ALSO"
.BR lpr (1),
.BR lprm (1),
.BR lpc (8),
.BR lpd (8)
.SH DIAGNOSTICS
.TP
.IB printer " is ready and printing"
The
.B lpq
program checks to see if there is a printer daemon.
If the daemon is hung, the super-user can
abort the current daemon and start a new one using
.BR lpc (8).
.TP
.BI "Waiting for " printer " to become ready (offline ?)"
The daemon could not open the printer device.  The printer may be
turned off-line.  This message can also occur if a printer is out
of paper, the paper is jammed, and so on.  Another possible cause is
that a process, such as an output filter, has exclusive use of the
device.  The only recourse in this case is to kill the offending
process and restart the printer with
.BR lpc .
.TP
.BI "waiting for " host " to come up"
A daemon is trying to connect to the remote machine named
.IR host ,
in order to send the files in the local queue.
If the remote machine is up,
.B lpd
on the remote machine is probably dead or
hung and should be restarted using
.IR lpc .
.br
.ne 5
.TP
.BI "sending to " host
The files are being transferred to the remote
.IR host ,
or else the local daemon has hung while trying to transfer the
files.
.TP
.BI "Warning: " printer " is down"
The printer has been marked as being unavailable with
.BR lpc .
.TP
.B Warning: no daemon present
The
.B lpd
process
overseeing
the spooling queue, as indicated in the ``lock'' file
in that directory, does not exist.  This normally occurs
only when the daemon has unexpectedly died.  Check the printer's
error log for a diagnostic from the deceased process; you can
restart the printer daemon with
.BR lpc .
.SH BUGS
.LP
.B lpq
may report unreliably.  The status as
reported may
not always reflect the actual state of the printer.
Under some circumstances,
.B lpq
reports that a printer is ready and printing when the daemon is,
in fact, hung.
.LP
Output formatting is sensitive to the line length of the terminal;
this can result in widely-spaced columns.
.LP
.B lpq
is sometimes unable to open various files when the
lock file is malformed.


ucb/lpr/lpq.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static char sccsid[] = "@(#)lpq.c 1.1 92/07/30 SMI"; /* from UCB 5.2 11/17/85 */
#endif not lint

/*
 * Spool Queue examination program
 *
 * lpq [+[n]] [-Pprinter] [user...] [job...]
 *
 * + means continually scan queue until empty
 * -P used to identify printer as per lpr/lprm
 */

#include "lp.h"

#include <sys/ioctl.h>

char	*user[MAXUSERS];	/* users to process */
int	users;			/* # of users in user array */
int	requ[MAXREQUESTS];	/* job number of spool entries */
int	requests;		/* # of spool requests */

static int	repeat;		/* + flag indicator */
static int	slptime = 30;	/* pause between screen refereshes */
static int	lflag;		/* long output option */

/*
 * Termcap stuff for fancy display
 */
#ifdef TERMCAP
struct sgttyb sbuf;
static unsigned ospeed;
static int	dumb;		/* whether to use capabilities */
static char	PC;		/* pad character for output */
static char	*UP;		/* up one line */
static char	*BC;		/* backspace character, other than \b */
static char	*CM;		/* cursor motion */
static char	*CL;		/* clear display */
static char	*TI;		/* terminal init for CM */
static char	*TE;		/* terminal clear for CM */
static char	*SO;		/* stand out start */
static char	*SE;		/* stand out end */

char	*tgetstr();
int	putch();		/* for tputs' */
#endif

main(argc, argv)
	char *argv[];
{
	register char *arg;
	register int n;

	name = argv[0];
	gethostname(host, sizeof host);
	openlog("lpd", 0, LOG_LPR);

	while (--argc) {
		if ((arg = *++argv)[0] == '+') {
			if (arg[1] != '\0')
				if ((n = atoi(&arg[1])) > 0)
					slptime = n;
			repeat++;
		} else if (arg[0] == '-')
			switch (arg[1]) {
			case 'P':		/* printer name */
				if (arg[2])
					printer = &arg[2];
				else if (argc > 1) {
					argc--;
					printer = *++argv;
				}
				break;

			case 'l':		/* long output */
				lflag++;
				break;

			default:
				usage();
		} else {
			if (isdigit(arg[0])) {
				if (requests >= MAXREQUESTS)
					fatal("too many requests");
				requ[requests++] = atoi(arg);
			} else {
				if (users >= MAXUSERS)
					fatal("too many users");
				user[users++] = arg;
			}
		}
	}
	if (printer == NULL && (printer = getenv("PRINTER")) == NULL)
		printer = DEFLP;
#ifdef TERMCAP
	dumb = termcap();
#endif

	if (repeat) {
#ifdef TERMCAP
		if (TI)
			tputs(TI, 0, putch);
#endif
		do {
#ifdef TERMCAP
			if (!dumb) {
				tputs(CL, 0, putch);
				tputs(tgoto(CM, 0, 0), 0, putch);
			}
#endif
			if ((n = displayq(lflag)) > 0)
				sleep(slptime);
		} while (n > 0);
#ifdef TERMCAP
		if (!dumb) {
			standout(stdout, "Hit return to continue");
			while (getchar() != '\n');
			if (TE)
				tputs(TE, 0, putch);
		}
#endif
	} else
		displayq(lflag);
	exit(0);
	/* NOTREACHED */
}

static
usage()
{
	printf("usage: lpq [-Pprinter] [-l] [+[n]] [user...] [job...]\n");
	exit(1);
}

/*
 * If we have the capability, print this in standout mode
 */
static
standout(f, s, a1, a2)
	FILE *f;
	char *s;
{
#ifdef TERMCAP
	if (SO)
		tputs(SO, 0, putch);
	fprintf(f, s, a1, a2);
	if (SO && SE)
		tputs(SE, 0, putch);
#else
	fprintf(f, s, a1, a2);
#endif
}

#ifdef TERMCAP
static char *
capstrings[] = {
	"bc", "cl", "cm", "so", "se", "ti", "te", "up",
	0
};

static char **
caps[] = {
	&BC, &CL, &CM, &SO, &SE, &TI, &TE, &UP,
};

/*
 * All we need from termcap is to clear screen and
 *   position cursor at the top; if these aren't available
 *   we say the terminal is dumb and let things scroll
 */
static
termcap()
{
	char *term, tbuf[BUFSIZ];
	static char buf[BUFSIZ/2];
	register short columns;
	char *bp = buf;
	register char **p, ***q, *cp;

	ioctl(0, TIOCGETP, (char *)&sbuf);
	ospeed = sbuf.sg_ospeed;
	if ((term = getenv("TERM")) != NULL && tgetent(tbuf, term) > 0) {
		for (p = capstrings, q = caps; *p != NULL; p++, q++)
			**q = tgetstr(*p, &bp);
		if ((cp = tgetstr("pc", &bp)) != NULL)
			PC = *cp;
	}
	return (CL == NULL || CM == NULL);
}

/*
 * Putchar writearound for tputs
 */
static
putch(c)
	char c;
{
	putchar(c);
}
#endif
